### PR TITLE
docs: add version locking document

### PR DESCRIPTION
**Pull request by tripodsan**

### CODE REVIEW

File: docs/version-locking.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+Ultimately, redering a specific set of services is initiated by the VCL via strains. By selecting the specific strain, the respective `x-ow-version-lock` header is sent to the respective `helix-dispatch` service. The strain is selected, either via the helix pages branch logic, or a strain condition, or a pre-flight condition.
+
+The _deploy_ script in the helix-pages repository ensures that the `versions.yaml` of the respective branch is properly transformed into a _strain_ in the `helix-config.yaml`.
+If the `versions.yaml` also contains an entry for the `pipeline`, it updates the `package.json`. 
+
+(the publishing step will update the respective edge dictionary for that strain).
+
+### Missing features
+
+- [ ] The `ow-x-version-lock` header must be strain dependant
```
Comment (trieloff) [inline docs/version-locking.md:70]:
```suggestion
- [ ] The `ow-x-version-lock` header must be strain dependent
```

File: docs/version-locking.md [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+- [ ] The `dispatch` version must be strain dependant
```
Comment (trieloff) [inline docs/version-locking.md:71]:
```suggestion
- [ ] The `dispatch` version must be strain dependent
```

File: docs/version-locking.md [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
```
Comment (trieloff) [inline docs/version-locking.md:72]:
```suggestion

- [ ] all other services that are called directly from the VCL must have strain dependent versions:
  - [ ] `purge` https://github.com/adobe/helix-publish/blob/c8eeba104aed31ba8107d5d689d8c63d8546933b/layouts/fastly/helix.vcl#L626
  - [ ] `query-index` https://github.com/adobe/helix-publish/blob/c8eeba104aed31ba8107d5d689d8c63d8546933b/layouts/fastly/helix.vcl#L711
  - [ ] `content-proxy` https://github.com/adobe/helix-publish/blob/c8eeba104aed31ba8107d5d689d8c63d8546933b/layouts/fastly/helix.vcl#L1130
 
```

File: docs/version-locking.md [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+1. helix creates the `breaking-december` version set that contains backward incompatible changes.
+2. project creates a `breaking-december` branch and (automatically) a `helix-versions.txt` (containing `breaking-december`). 
+3. project updates its client code until the breaking issues are fixed.
+4. project merges its `breaking-december` branch back to `main`, which now contains the `helix-versions.txt` with the `breaking-december` version name.
+5. helix creates the `breaking-january` version set
+6. project **doesn't update** to breaking-january
+7. helix creates the `breaking-february` version set
+8. `breaking-december` will now be the new default and all services are updated accordingly.
+
+And so on... As long a 1 project still contains a reference to an old _version_, the strain in helix-pages needs to remain. The risk here is to reach the limit of the edge dictionaries or to loose oversight.
```
Comment (trieloff) [inline docs/version-locking.md:100]:
```suggestion
And so on… As long a 1 project still contains a reference to an old _version_, the strain in helix-pages needs to remain. The risk here is to reach the limit of the edge dictionaries or to loose oversight.
```

Comment (trieloff) [reply docs/version-locking.md:100]: I think a bigger risk is the customers who want to stay perpetually on an old version, but also have new features. Therefore I'd treat `version.txt` as an indication of preference, and not as a hard choice.

**Review by trieloff — APPROVED** [review]

File: docs/version-locking.md [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+- [ ] The `helix-version-picker` https://github.com/adobe/helix-home/issues/167
+- 
+
+
+## Developer Experience (customer)
+
+Ideally the developer just uses git and the helix command line to:
+
+1. checkout his project
+2. running `hlx` tells him that there are new upcomming breaking changes
```
Comment (rofe) [inline docs/version-locking.md:119]: upcoming

File: docs/version-locking.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+3. he can choose to automatically _start testing_ the new changes. the cli would create the respective branch, create the `helix-versions.txt`, update helix pages
```
Comment (rofe) [inline docs/version-locking.md:120]: he -> they

File: docs/version-locking.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+- the VCL deployed via helix-publish to fastly (inner-cdn config)
+- the project code (helix-query.yaml, head.html, scripts, css, etc...)
+
+A helix pages _version_ or _feature_ is defined by:
+
+- an identifier (eg. breaking-december)
+- a set of services with specific versions (eg. word2md v20201201.1, content-proxy v20201201.4)
+- a helix pages code branch
+
+The version set definition is recorded in a `versions.yaml` which is stored in the respective helix pages code branch. For example, a `breaking-december` _version_ would have `versions.yaml` in the `breaking-december` branch in the helix-pages repo:
```
Comment (rofe) [inline docs/version-locking.md:30]: Should we call it `helix-versions.yaml` to align with other Helix config files?

**Review by rofe — COMMENTED** [review]
Sorry for being late to the party...

I second @trieloff's point that the customer's version lock file should be treated as a preference that will be ignored past a certain point in time, as we continually retire older sets (expiry rules tbd). It should also be made clear that old version sets won't be patched (with the only exception of a major security vulnerabilities). These rules are here to minimize the Helix's team's efforts to maintain version sets, and set customer expectations that:
- their journey on a version lock is only temporary in nature
- they don't get to stay there indefinitely (think AEM on-prem customers expecting bug fixes and even "feature packs" for 5 year-old versions)
- they should have a clear plan for bringing their codebase back to `main` as soon as the `breaking-<younameit>` branch in `helix-pages` is merged
If we don't set a clear policy around this from the beginning, I'm worried that we will end up having to support an unlimited number of version sets and backport fixes and even features to customers who are "stuck" (for whatever reason) on old version sets.

The CLI could inform developers if their selected version set is about to expire or no longer available.

**Comment by kptdobe** [comment]
You keep referring to the cli but to use helix-pages, the cli is an "advanced" use case. You can do everything from a GitHub repo without the need of using the cli. If you look all our existing "customers", I wonder how many use (even know about) the cli. I even wonder if @davidnuescheler uses it...
If this is the case, all those helix-pages consumers expect their site to still be running whatever we change, especially on the DOM side.
Why should I do something so complicated if I have one single css line in my `style.css` override and we are going to break it because we change the DOM ?
If helix-pages target audience is non technical people or people that can just hack a css quickly, the only way to orchestrate this is to do it for them via PR we make on their repos.

File: docs/version-locking.md (hunk 7 above)
Comment (tripodsan) [reply docs/version-locking.md:30]: right now, I think its only something used in the helix-pages project and not used by customers... but yes...maybe.

**Comment by tripodsan** [comment]
currently, developing w/o helix cli is not possible. reloading the live website is so slow, that it's not possible to develop css and js w/o cli.

creating automated PRs would be an option....we would need to fork every customers repo and then create the PRs - or have them the helix-github-bot installed, with enough permissions to write to their repo. but we at least could open issues on all the repositories we know, informing the customer that they should check and update their project.

**Pull request merged**